### DISCUSSION TITLE
"this.fabric_eventDispatcher" is null

### DISCUSSION THREAD

**Discussion by quiqueck** [FabricMC/fabric-api#3214] [Mod Dev Support] [answered]
### Description

Hi.

I tried to compile my Mods (BetterEnd) against the latest 0.86 of FAPI. This introduces a lot of errors when starting the game that all look like this:
> Unable to load model: 'fabric:tag/item/shears' referenced from: fabric:tag/item/shears: java.lang.NullPointerException: Cannot invoke "net.fabricmc.fabric.impl.client.model.loading.ModelLoadingEventDispatcher.loadModel(net.minecraft.resources.ResourceLocation)" because "this.fabric_eventDispatcher" is null

I did read about the changes related to Model Loading (https://github.com/FabricMC/fabric/pull/3145), but did not yet implement those. Is this error an expected result of those changes?

The Tag mentioned in the log is loaded from a Folder generated by a DataGenerator.

**Comment by quiqueck** [comment]
There is also another, more subtle issue with 0.86 here:
https://github.com/quiqueck/BetterEnd/issues/273

That looks like the content of a custom registry was not properly loaded, and is also introduced by FAPI 0.86, but it is probably unrelated to the above issue...

**Comment by PepperCode1** [comment]
The only other case of the null event dispatcher that I have seen is when Model Loading API v1 was used with a Models v0 version older than 0.4.0. This is not a working or supported configuration. Check the module versions in the log during start up to see whether your issue has the same cause.

**Comment by quiqueck** [comment]
Ah, thanks. Now I have at least a starting point :)

>	- fabric-model-loading-api-v1 1.0.1+9386d8a777
>	- fabric-models-v0 0.3.35+b3afc78b27

**Comment by quiqueck** [comment]
ok, so this is probably a dumb question. But how do I get a Version 0.4.0 or higher for the models v0 API?

And what will happen if the mod is loaded with other mods that still use fabric-models-v0 0.3.35? Will this crash?

**Reply by PepperCode1** [reply]
Fabric API bundles Models v0 0.4.0. If a different version is being loaded in your development environment, your build script needs to be fixed. It is possible for a declared mod dependency to pull in an old version of Models v0.

Mods should not be depending directly on Models v0 0.3.35. The API of Models v0 did not change between 0.3.35 and 0.4.0, so mods that use it in a supported way (meaning they do not interact with the module implementation) will continue working and will not care that Models v0 has been updated. There will be no crashes.

**Comment by quiqueck** [comment]
Hm. For me the issue appears when I add the latest Version of EMI to my testing environment.

If I use `modLocalRuntime "dev.emi:emi-fabric:${emi_version}"`, the old v0 0.3.35 is loaded in my dev environment, which results in the above error.

If I drop it from he build files, and just add it to my *run/mods* folder, the game will crash and complain that `net.fabricmc.fabric.api.client.model.ModelLoadingRegistry` was not found.

I noticed, that the models api is not loaded at all in this configuration. As far as I am aware, I am not doing anything special in my build scripts...

**Reply by PepperCode1** [reply]
Disable EMI's transitive Fabric API dependency, then.

**Comment by quiqueck** [comment]
Maybe I am not seeing something really obvious here...

Disabling transitive dependencies should be the same as putting the mod in the `run/mods` folder, right?

When I add EMI like this:
```
modLocalRuntime("dev.emi:emi-fabric:${emi_version}") {
  transitive = false;
}
```

neither the old v0 0.3.35 nor any newer version is loaded (same as when I manually put the mod in `run/mods`), and thus the game crashes with
> java.lang.ClassNotFoundException: net.fabricmc.fabric.api.client.model.ModelLoadingRegistry

**Reply by PepperCode1** [reply]
The standard Fabric API Gradle dependency does not include deprecated modules.

**Comment by quiqueck** [comment]
I am really sorry, but I do not get the version handling here...

I disabled EMI, and just compiled  against 0.86 without changing any code. This will not include *fabric-model-loading-api-v1* (as expected), but it will bring in *fabric-models-v0 0.3.35*. Shouldn't that use the new 0.4.x of fabric-models-v0?

If I use this non upgraded version in another 0.68 Mod (that is also using the *fabric-models-v0* API) as dependency, the game will crash as *fabric-models-v0* is not loaded at all (Both mods unaltered, I just changed the FAPI version from 0.85 to 0.86).

If I update my code to use the new API, *fabric-model-loading-api-v1* is included and *fabric-models-v0* is not. However, this will result in a crash when EMI (or I suspect any mod that uses the old API) is added to the client.

**Reply by PepperCode1** [reply, marked as answer]
Try re-enabling transitive dependencies and adding the deprecated modules Fabric API dependency. The Fabric API 0.86.0 dependency does not have Models v0 at all, so the latest version that is available in your environment is 0.3.35, provided by EMI. The deprecated modules dependency will provide the 0.4.0 version, which should override the 0.3.35 version. Note that production Fabric API jars automatically include deprecated modules to allow mods that depend on them to continue working.

**Comment by quiqueck** [comment]
Ah, that makes sense. Thanks a lot for taking all the time. I really appreciate the help :)
